orchagent/nhgorch.h

```diff
@@ -0,0 +1,242 @@
+        return m_nh_key.to_string() +
+                ", SAI ID: " + std::to_string(m_gm_id);
+    }
+
+private:
+    /* The key of the next hop of this member. */
+    NextHopKey m_nh_key;
+
+    /* The group member SAI ID for this member. */
+    sai_object_id_t m_gm_id;
+};
+
+/* Map indexed by NextHopKey, containing the SAI ID of the group member. */
+typedef std::map<NextHopKey, NextHopGroupMember> NhgMembers;
+
+/*
+ * NextHopGroup class representing a next hop group object.
+ */
+class NextHopGroup
+{
+public:
+    /* Constructors. */
+    explicit NextHopGroup(const NextHopGroupKey& key);
+    NextHopGroup(NextHopGroup&& nhg);
+    NextHopGroup& operator=(NextHopGroup&& nhg);
+
+    /* Destructor. */
+    virtual ~NextHopGroup() { desync(); }
+
+    /* Sync the group, creating the group's and members SAI IDs. */
+    bool sync();
+
+    /* Desync the group, reseting the group's and members SAI IDs.  */
+    bool desync();
+
+    /*
+     * Update the group based on a new next hop group key.  This will also
+     * perform any sync / desync necessary.
+     */
+    bool update(const NextHopGroupKey& nhg_key);
+
+    /* Check if the group contains the given next hop. */
+    inline bool hasNextHop(const NextHopKey& nh_key) const {
+                            return m_members.find(nh_key) != m_members.end(); }
+
+    /* Validate a next hop in the group, syncing it. */
+    bool validateNextHop(const NextHopKey& nh_key);
+
+    /* Invalidate a next hop in the group, desyncing it. */
+    bool invalidateNextHop(const NextHopKey& nh_key);
+
+    /* Increment the number of existing groups. */
+    static inline void incCount() { ++m_count; }
+
+    /* Decrement the number of existing groups. */
+    static inline void decCount() { assert(m_count > 0); --m_count; }
+
+    /* Getters / Setters. */
+    inline const NextHopGroupKey& getKey() const { return m_key; }
+    inline sai_object_id_t getId() const { return m_id; }
+    static inline unsigned int getCount() { return m_count; }
+    inline bool isTemp() const { return m_is_temp; }
+    inline void setTemp(bool is_temp) { m_is_temp = is_temp; }
+    inline bool isSynced() const { return m_id != SAI_NULL_OBJECT_ID; }
+    inline size_t getSize() const { return m_members.size(); }
+
+    /* Convert NHG's details to a string. */
+    std::string to_string() const
+    {
+        return m_key.to_string() + ", SAI ID: " + std::to_string(m_id);
+    }
+
+private:
+
+    /* The next hop group key of this group. */
+    NextHopGroupKey m_key;
+
+    /* The SAI ID of the group. */
+    sai_object_id_t m_id;
+
+    /* Members of this next hop group. */
+    NhgMembers m_members;
+
+    /* Whether the group is temporary or not. */
+    bool m_is_temp;
+
+    /*
+     * Number of existing groups.  Incremented when an object is created and
+     * decremented when an object is destroyed.  This will also account for the
+     * groups created by RouteOrch.
+     */
+    static unsigned int m_count;
```

Comment:
why maintain it here? Why not NghOrch? Does groups created by routeOrch take a different path?

Reply:
Not sure what this comment means - this is nhgOrch!

Reply:
I meant why inside the class. Either is fine.